src/libstd/ffi/c_str.rs

```diff
@@ -255,6 +255,34 @@ impl CString {
         Box::into_raw(self.inner) as *mut libc::c_char
     }
 
+    /// Converts the `CString` into a `String` if it contains valid Unicode data.
+    ///
+    /// On failure, ownership of the original `CString` is returned.
+    #[unstable(feature = "convert", reason = "recently added", issue = "27704")]
+    pub fn into_string(self) -> Result<String, CString> {
+        String::from_utf8(self.into_bytes())
+            .map_err(|e| unsafe { CString::from_vec_unchecked(e.into_bytes()) })
+    }
+
+    /// Returns the underlying byte buffer.
+    ///
+    /// The returned buffer does **not** contain the trailing nul separator and
+    /// it is guaranteed to not have any interior nul bytes.
+    #[unstable(feature = "convert", reason = "recently added", issue = "27704")]
+    pub fn into_bytes(self) -> Vec<u8> {
```

Comment:
@arcnmx In general the C string type is considered a "container of non-nul bytes" in the sense that when you take a look at the innards (e.g. `as_bytes` and friends) you don't get the trailing nul by default.

It's also currently idiomatic to have both the `Into` implementation as well as the `into_foo` concrete method to call itself, so I don't think we'd want to remove this in favor of *only* an `Into` implementation.